File: worker/agents/orchestrator.ts
import type Redis from "ioredis";
import { runSkillAgent } from "../lib/sdk";
import { makeBus } from "../lib/bus";
import { runExplore } from "./curator";
import { runRebuildPath } from "./learningPath";
import { runNarrate } from "./voice";

export interface OrchestrateJob {
  uid: string;
  jobId: string;
  request: string;
}

interface Plan {
  action: "explore" | "rebuild_path" | "narrate" | "answer";
  topic?: string;
  text?: string;
  answer?: string;
}

const ROUTER_PROMPT = (request: string) => `You are Tiruno's orchestrator. Classify the user's request and decide which agent should handle it. Respond with ONLY a JSON object (no prose, no fences):
{"action":"explore|rebuild_path|narrate|answer","topic":"<for explore>","text":"<for narrate>","answer":"<for answer: a short helpful reply>"}

Rules:
- "research X", "what's new in X", "find me / get me X", "trending X" → action "explore", topic = X.
- "what should I learn", "build/rebuild my path", "plan my studies" → action "rebuild_path".
- "read/narrate this", "say it out loud" → action "narrate", text = the thing to narrate.
- Anything else (greetings, general questions) → action "answer" with a concise answer.

User request: "${request}"`;

function parsePlan(text: string): Plan {
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/i);
  const c = fenced ? fenced[1] : text;
  const start = c.indexOf("{");
  const end = c.lastIndexOf("}");
  if (start !== -1 && end !== -1) {
    try {
      const o = JSON.parse(c.slice(start, end + 1));
      if (o && typeof o.action === "string") return o as Plan;
    } catch {
      /* fall through */
    }
  }
  return { action: "answer", answer: text.trim().slice(0, 400) };
}

/**
 * Orchestrator agent (LLM): the conductor for fuzzy / voice requests like
 * "Bear, what should I do next?" — decides the plan and delegates to worker agents.
 */
export async function runOrchestrate(redis: Redis, job: OrchestrateJob): Promise<void> {
  const bus = makeBus(redis);
  await bus.publish(job.uid, { jobId: job.jobId, type: "progress", status: "researching", step: "Thinking about what you need…" });

  const res = await runSkillAgent({ prompt: ROUTER_PROMPT(job.request), skills: [], maxTurns: 3 });
  const plan = res.ok ? parsePlan(res.text) : { action: "answer" as const, answer: "I couldn't process that — try rephrasing." };

  switch (plan.action) {
    case "explore":
      await runExplore(redis, { uid: job.uid, jobId: job.jobId, topic: plan.topic || job.request });
      return;
    case "rebuild_path":
      await runRebuildPath(redis, { uid: job.uid, jobId: job.jobId });
      return;
    case "narrate":
      await runNarrate(redis, { uid: job.uid, jobId: job.jobId, text: plan.text || job.request });
      return;
    default:
      await bus.publish(job.uid, {
        jobId: job.jobId,
        type: "done",
        status: "ready",
        result: { answer: plan.answer ?? "Done." },
      });
  }
}

File: worker_py/agents/orchestrator.py
"""Orchestrator agent. Port of ``worker/agents/orchestrator.ts``.

The conductor for fuzzy / voice requests ("Bear, what should I do next?") —
classifies the request and delegates to a worker agent.
"""

import json
import re
from typing import Any, Dict

from redis.asyncio import Redis

from ..agent import run_skill_agent
from ..bus import publish as bus_publish
from .curator import run_explore
from .learning_path import run_rebuild_path
from .voice import run_narrate


def _router_prompt(request: str) -> str:
    return (
        "You are Tiruno's orchestrator. Classify the user's request and decide which agent should handle it. "
        "Respond with ONLY a JSON object (no prose, no fences):\n"
        '{"action":"explore|rebuild_path|narrate|answer","topic":"<for explore>","text":"<for narrate>",'
        '"answer":"<for answer: a short helpful reply>"}\n\n'
        "Rules:\n"
        '- "research X", "what\'s new in X", "find me / get me X", "trending X" → action "explore", topic = X.\n'
        '- "what should I learn", "build/rebuild my path", "plan my studies" → action "rebuild_path".\n'
        '- "read/narrate this", "say it out loud" → action "narrate", text = the thing to narrate.\n'
        '- Anything else (greetings, general questions) → action "answer" with a concise answer.\n\n'
        f'User request: "{request}"'
    )


def _parse_plan(text: str) -> Dict[str, Any]:
    m = re.search(r"```(?:json)?\s*([\s\S]*?)```", text or "", re.IGNORECASE)
    c = m.group(1) if m else (text or "")
    start = c.find("{")
    end = c.rfind("}")
    if start != -1 and end != -1:
        try:
            o = json.loads(c[start : end + 1])
            if isinstance(o, dict) and isinstance(o.get("action"), str):
                return o
        except Exception:  # noqa: BLE001
            pass
    return {"action": "answer", "answer": (text or "").strip()[:400]}


async def run_orchestrate(redis: Redis, job: Dict[str, str]) -> None:
    uid, job_id, request = job["uid"], job["jobId"], job["request"]
    await bus_publish(redis, uid, {"jobId": job_id, "type": "progress", "status": "researching", "step": "Thinking about what you need…"})

    res = await run_skill_agent(prompt=_router_prompt(request), skills=[], max_turns=3)
    plan = _parse_plan(res.text) if res.ok else {"action": "answer", "answer": "I couldn't process that — try rephrasing."}

    action = plan.get("action")
    if action == "explore":
        await run_explore(redis, {"uid": uid, "jobId": job_id, "topic": plan.get("topic") or request})
    elif action == "rebuild_path":
        await run_rebuild_path(redis, {"uid": uid, "jobId": job_id})
    elif action == "narrate":
        await run_narrate(redis, {"uid": uid, "jobId": job_id, "text": plan.get("text") or request})
    else:
        await bus_publish(redis, uid, {"jobId": job_id, "type": "done", "status": "ready", "result": {"answer": plan.get("answer") or "Done."}})
